File: Data Structures & Algorithms/binary-tree-diameter/submission-0.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    max_dia = 0
    def diameterOfBinaryTree(self, root: Optional[TreeNode]) -> int:
        
        if root is None:
            return 0
        self.dfs(root)
        return self.max_dia

    def dfs(self,node):
        if node is None:
            return 0
        left_height = self.dfs(node.left)
        right_height = self.dfs(node.right)
        diameter = left_height+right_height
        self.max_dia = max(self.max_dia, diameter)
        return 1 + max(left_height, right_height)
        '''
dfs(node):

    visit root node
    left_height = dfs(left)
    right_hright = dfs(right)
    return 1+ max{left_height, right_height}
        '''

File: Data Structures & Algorithms/balanced-binary-tree/submission-0.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    boolean valid = true;
    public boolean isBalanced(TreeNode root) {
        dfs(root);
        return this.valid;
    }
    public int dfs(TreeNode root){
        if (root == null){
            return 0;
        }
        int left_height = dfs(root.left);
        int right_height = dfs(root.right);
        if(Math.abs(left_height - right_height)>1){
            this.valid = false;
        }

        return 1 + Math.max(left_height, right_height);
    }
}

/*

valid

dfs(root):
    if(root = None):
        retun 0
    print(root.val)
    left_height = dfs(root.left)
    right_height = dfs(root.right)
    if((left_height - right_height) >1):
        self.valid = False
    return 1+ max(left_height,right_height)

*/